File: api/schemas/study_anki.py
"""Strict, metadata-only HTTP contracts for Study Anki portability."""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    StrictBool,
    StrictInt,
    StrictStr,
    field_validator,
)


class _Strict(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)


def _visible(value: str, *, maximum: int, field_name: str) -> str:
    if value != value.strip() or not value or len(value) > maximum:
        raise ValueError(f"{field_name} must be bounded visible text")
    if any(ord(char) < 32 or ord(char) == 127 for char in value):
        raise ValueError(f"{field_name} must be bounded visible text")
    return value


class AnkiHttpOptions(_Strict):
    schema_version: Literal[1] = 1
    syllabus_unit_id: StrictStr | None = Field(default=None, max_length=64)
    deck_names: tuple[StrictStr, ...] = Field(default_factory=tuple, max_length=100)

    @field_validator("syllabus_unit_id")
    @classmethod
    def unit_id_is_safe(cls, value: str | None) -> str | None:
        if value is not None:
            _visible(value, maximum=64, field_name="syllabus_unit_id")
            if not value[0].islower() and not value[0].isdigit():
                raise ValueError("invalid syllabus_unit_id")
            if any(char not in "abcdefghijklmnopqrstuvwxyz0123456789_-" for char in value):
                raise ValueError("invalid syllabus_unit_id")
        return value

    @field_validator("deck_names", mode="before")
    @classmethod
    def tuple_deck_names(cls, value: object) -> object:
        return tuple(value) if isinstance(value, list) else value

    @field_validator("deck_names")
    @classmethod
    def deck_names_are_safe(cls, values: tuple[str, ...]) -> tuple[str, ...]:
        if len(set(values)) != len(values):
            raise ValueError("duplicate deck_names")
        for value in values:
            _visible(value, maximum=200, field_name="deck_name")
        return values


class AnkiImportPublishRequest(_Strict):
    upload_id: StrictStr = Field(min_length=1, max_length=128)
    request_id: StrictStr = Field(min_length=1, max_length=256)
    options: AnkiHttpOptions = Field(default_factory=AnkiHttpOptions)

    @field_validator("upload_id", "request_id")
    @classmethod
    def ids_are_safe(cls, value: str, info: object) -> str:
        return _visible(value, maximum=256, field_name=str(getattr(info, "field_name", "id")))


class AnkiImportPreviewResponse(_Strict):
    schema_version: Literal[1] = 1
    job_id: StrictStr = Field(min_length=1, max_length=128)
    status: Literal["preview_ready", "processing", "failed", "published"]
    card_count: StrictInt = Field(ge=0, le=10_000)
    transformed_count: StrictInt = Field(ge=0, le=10_000)
    skipped_count: StrictInt = Field(ge=0, le=10_000)
    rejected_count: StrictInt = Field(ge=0, le=10_000)
    package_sha256: str = Field(pattern=r"^[0-9a-f]{64}$")
    collection_member: Literal["collection.anki2", "collection.anki21"]
    message: str | None = Field(default=None, max_length=200)


class AnkiImportStatusResponse(AnkiImportPreviewResponse):
    receipt_id: str | None = Field(default=None, max_length=512)


class AnkiCompatibilityReceiptResponse(_Strict):
    schema_version: Literal[1] = 1
    receipt_id: str = Field(min_length=1, max_length=512)
    plan_id: str = Field(min_length=1, max_length=512)
    request_id: str = Field(min_length=1, max_length=256)
    payload_sha256: str = Field(pattern=r"^[0-9a-f]{64}$")
    package_sha256: str = Field(pattern=r"^[0-9a-f]{64}$")
    collection_sha256: str = Field(pattern=r"^[0-9a-f]{64}$")
    collection_member: Literal["collection.anki2", "collection.anki21"]
    card_count: StrictInt = Field(ge=0, le=10_000)
    transformed_count: StrictInt = Field(ge=0, le=10_000)
    skipped_count: StrictInt = Field(ge=0, le=10_000)
    card_ids: tuple[str, ...] = Field(max_length=10_000)
    deck_names: tuple[str, ...] = Field(max_length=1_000)
    tags: tuple[str, ...] = Field(max_length=1_000)
    media_names: tuple[str, ...] = Field(max_length=500)
    syllabus_unit_id: str | None = Field(default=None, max_length=64)
    created_at: datetime


class AnkiImportPublishResponse(_Strict):
    schema_version: Literal[1] = 1
    status: Literal["published", "replayed"]
    receipt: AnkiCompatibilityReceiptResponse


class AnkiExportRequest(_Strict):
    schema_version: Literal[1] = 1
    options: AnkiHttpOptions = Field(default_factory=AnkiHttpOptions)


class AnkiExportReceipt(_Strict):
    schema_version: Literal[1] = 1
    receipt_id: str = Field(min_length=1, max_length=512)
    plan_id: str = Field(min_length=1, max_length=512)
    plan_revision: StrictInt = Field(ge=1)
    syllabus_version: StrictInt = Field(ge=1)
    package_sha256: str = Field(pattern=r"^[0-9a-f]{64}$")
    card_count: StrictInt = Field(ge=0, le=10_000)
    stable_note_guids: tuple[str, ...] = Field(max_length=10_000)
    stable_model_ids: tuple[StrictInt, ...] = Field(max_length=16)
    stable_deck_ids: tuple[StrictInt, ...] = Field(max_length=1_000)
    created_at: datetime


class AnkiExportResponse(_Strict):
    schema_version: Literal[1] = 1
    download_id: str = Field(min_length=1, max_length=128)
    receipt: AnkiExportReceipt


__all__ = [
    "AnkiCompatibilityReceiptResponse",
    "AnkiExportReceipt",
    "AnkiExportRequest",
    "AnkiExportResponse",
    "AnkiHttpOptions",
    "AnkiImportPreviewResponse",
    "AnkiImportPublishRequest",
    "AnkiImportPublishResponse",
    "AnkiImportStatusResponse",
]

File: frontend/src/lib/types/study-anki.ts
import { z } from 'zod'

const boundedId = z.string().min(1).max(512).refine((value) => value === value.trim() && !/[\u0000-\u001f\u007f]/.test(value))
const sha256 = z.string().regex(/^[0-9a-f]{64}$/)

export const ankiHttpOptionsSchema = z.object({
  schema_version: z.literal(1).default(1),
  syllabus_unit_id: z.string().regex(/^[a-z0-9][a-z0-9_-]{0,63}$/).nullable().optional(),
  deck_names: z.array(z.string().min(1).max(200)).max(100).default([]),
}).strict()

export const ankiImportPreviewSchema = z.object({
  schema_version: z.literal(1),
  job_id: z.string().regex(/^anki_job:[a-f0-9]{32,64}$/),
  status: z.enum(['preview_ready', 'processing', 'failed', 'published']),
  card_count: z.number().int().min(0).max(10_000),
  transformed_count: z.number().int().min(0).max(10_000),
  skipped_count: z.number().int().min(0).max(10_000),
  rejected_count: z.number().int().min(0).max(10_000),
  package_sha256: sha256,
  collection_member: z.enum(['collection.anki2', 'collection.anki21']),
  message: z.string().max(200).nullable().optional(),
}).strict()

export const ankiImportStatusSchema = ankiImportPreviewSchema.extend({
  receipt_id: z.string().max(512).nullable(),
}).strict()

export const ankiCompatibilityReceiptSchema = z.object({
  schema_version: z.literal(1),
  receipt_id: boundedId,
  plan_id: boundedId,
  request_id: boundedId,
  payload_sha256: sha256,
  package_sha256: sha256,
  collection_sha256: sha256,
  collection_member: z.enum(['collection.anki2', 'collection.anki21']),
  card_count: z.number().int().min(0).max(10_000),
  transformed_count: z.number().int().min(0).max(10_000),
  skipped_count: z.number().int().min(0).max(10_000),
  card_ids: z.array(boundedId).max(10_000),
  deck_names: z.array(z.string().min(1).max(200)).max(1_000),
  tags: z.array(z.string().min(1).max(128)).max(1_000),
  media_names: z.array(z.string().min(1).max(512)).max(500),
  syllabus_unit_id: z.string().regex(/^[a-z0-9][a-z0-9_-]{0,63}$/).nullable(),
  created_at: z.string().datetime({ offset: true }),
}).strict()

export const ankiImportPublishSchema = z.object({
  schema_version: z.literal(1),
  status: z.enum(['published', 'replayed']),
  receipt: ankiCompatibilityReceiptSchema,
}).strict()

export const ankiExportReceiptSchema = z.object({
  schema_version: z.literal(1),
  receipt_id: boundedId,
  plan_id: boundedId,
  plan_revision: z.number().int().min(1),
  syllabus_version: z.number().int().min(1),
  package_sha256: sha256,
  card_count: z.number().int().min(0).max(10_000),
  stable_note_guids: z.array(z.string().regex(/^[0-9a-f]{32}$/)).max(10_000),
  stable_model_ids: z.array(z.number().int().positive()).max(16),
  stable_deck_ids: z.array(z.number().int().positive()).max(1_000),
  created_at: z.string().datetime({ offset: true }),
}).strict()

export const ankiExportResponseSchema = z.object({
  schema_version: z.literal(1),
  download_id: z.string().regex(/^anki_download:[a-f0-9]{32,64}$/),
  receipt: ankiExportReceiptSchema,
}).strict()

export type AnkiHttpOptions = z.infer<typeof ankiHttpOptionsSchema>
export type AnkiImportPreview = z.infer<typeof ankiImportPreviewSchema>
export type AnkiImportStatus = z.infer<typeof ankiImportStatusSchema>
export type AnkiCompatibilityReceipt = z.infer<typeof ankiCompatibilityReceiptSchema>
export type AnkiImportPublish = z.infer<typeof ankiImportPublishSchema>
export type AnkiExportResponse = z.infer<typeof ankiExportResponseSchema>

function invalidResponse(): Error {
  return new Error('Invalid Study Anki response')
}

function decode<T>(schema: z.ZodType<T>, value: unknown): T {
  const parsed = schema.safeParse(value)
  if (!parsed.success) throw invalidResponse()
  return parsed.data
}

export const decodeAnkiImportPreview = (value: unknown) => decode(ankiImportPreviewSchema, value)
export const decodeAnkiImportStatus = (value: unknown) => decode(ankiImportStatusSchema, value)
export const decodeAnkiImportPublish = (value: unknown) => decode(ankiImportPublishSchema, value)
export const decodeAnkiExportResponse = (value: unknown) => decode(ankiExportResponseSchema, value)
